File: 백준/Gold/1339. 단어 수학/단어 수학.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        List<String> arr = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            arr.add(br.readLine());
        }

        Map<Character, Integer> alphabetMap = new HashMap<>();
        for (String s : arr) {
            int sLength = s.length();
            for (int i = 0; i < sLength; i++) {
                char c = s.charAt(i);
                int val = (int) Math.pow(10 ,sLength - i - 1);
                if (alphabetMap.containsKey(c)) {
                    val += alphabetMap.get(c);
                }
                alphabetMap.put(c, val);
            }
        }
        List<Map.Entry<Character, Integer>> lst = new ArrayList<>(alphabetMap.entrySet());
        lst.sort((a, b) -> b.getValue().compareTo(a.getValue()));

        Map<Character, Integer> alphaToNumber = new HashMap<>();
        int num = 9;
        for (Map.Entry<Character, Integer> entry : lst) {
            alphaToNumber.put(entry.getKey(), num);
            num--;
        }

        long answer = 0;
        for (String s: arr) {
            int sLength = s.length();
            for (int i = 0; i < sLength; i++) {
                char c = s.charAt(i);
                answer += (long) alphaToNumber.get(c) * (int) Math.pow(10, sLength - i - 1);
            }
        }
        System.out.println(answer);

    }
}

File: 백준/Gold/1339. 단어 수학/단어 수학.py
import sys
input = sys.stdin.readline

def solution():
    dic = {}

    N = int(input().strip())
    arr = []
    for _ in range(N):
        arr.append(input().strip())

    for i in range(N):
        w = arr[i]
        w_length = len(w)
        for j in range(w_length):
            dic[w[j]] = dic.get(w[j], 0) + 10**(w_length - j - 1)

    lst = [(val, key) for key, val in dic.items()]
    lst.sort(reverse=True)

    alphaToNumber = {}
    num = 9
    for val, key in lst:
        alphaToNumber[key] = num
        num -= 1

    answer = 0
    for i in range(N):
        w = arr[i]
        w_length = len(w)
        for j in range(w_length):
            answer += alphaToNumber[w[j]] * 10**(w_length - j - 1)
    print(answer)


solution()